crates/wasm-smith/src/lib.rs

```diff
@@ -1570,3 +1799,56 @@ impl AvailableAliases {
         });
     }
 }
+
+/// A helper structure to keep track of possible instantiations a module can do
+/// at any one point in time.
+///
+/// This structure is similar to `AvailableAliases` where it is used as part of
+/// the `arbitrary_instances` method to generate the module linking instance
+/// section. Candidates for instantiation are stored in `choices`, and if that
+/// array is empty then there are not possible candidates for instantiation at
+/// this time.
+///
+/// This will be dynamically updated over time with new choices as they become
+/// available (e.g. as imports are defined, aliases are defined, etc).
+#[derive(Default)]
+struct AvailableInstantiations {
+    choices: Vec<Instantiation>,
+}
+
+struct Instantiation {
+    /// The module index which can be instantiated.
+    module: u32,
+    /// The argument vector that must be provided to instantiation, where each
+    /// vector element here is a list of module items which match the import's
+    /// type.
+    ///
+    /// Each sub-vector here is guaranteed to have a length of at least 1, and
+    /// if this instantiation candidate is chosen we'll use fuzz input to select
+    /// which item is used for instantiation.
```

Comment:
Why is each sub-vector guaranteed to have a length of at least 1? What about modules with no imports? They shouldn't take any arguments, right?

Reply:
Ahhhhh because this isn't `[ set_of_complete_arguments, another_set_of_complete_arguments, ... ]` this is `[ options_for_first_arg, options_for_second_arg, ... ]`. Okay, this makes a lot of sense and avoids a combinatorial explosion.

Could we clarify this a little with something like:

> Each sub-vector `args[i]` contains all the valid options for the `i`th instantiation argument of our `module`. The entry is never empty, because that would mean that the module cannot be instantiated, in which case we do not construct this structure. In the following example, both `$g1` and `$g2` are possible arguments for `$nested.imported-global`, so the associated `args[i]` would be `[$g1, $g2]`
>
> ```wat
> (module
>   (module $nested
>     (import $imported-global "g" (mut i32)))
>   (global $g1 (mut i32) (i32.const 0))
>   (global $g2 (mut i32) (i32.const 42)))
> ```